### ISSUE TITLE
Build error on Ubuntu 19.10 with GCC

### ISSUE THREAD

**Issue by jtracey** [shadow/shadow#737]
**Describe the issue**
A clean build of master (and every release going back a ways) fails when doing a vanilla clean build when using GCC. Debug builds still complete fine (albeit with some compiler warnings), as do vanilla clang builds (with even more warnings). Key lines from the build fail output:
```
[ 98%] Building C object src/preload/CMakeFiles/shadow-interpose.dir/shd-interposer.c.o
In file included from /usr/include/stdio.h:867,
                 from /home/j3tracey/shadow/src/preload/shd-interposer.c:12:
/home/j3tracey/shadow/src/preload/shd-interposer.c: In function ‘fprintf’:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:98:1: error: inlining failed in call to always_inline ‘fprintf.localalias’: redefined extern inline functions are not considered for inlining
   98 | fprintf (FILE *__restrict __stream, const char *__restrict __fmt, ...)
      | ^~~~~~~
/home/j3tracey/shadow/src/preload/shd-interposer.c:59:9: note: called from here
   59 |         fprintf(stderr, "dlsym(%s): dlerror(): %s\n", funcstr, errorMessage); \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/j3tracey/shadow/src/preload/shd-interposer.c:69:9: note: in expansion of macro ‘SETSYM_OR_FAIL’
   69 |         SETSYM_OR_FAIL(director.next.func, #func); \
      |         ^~~~~~~~~~~~~~
/home/j3tracey/shadow/src/preload/shd-interposer.c:396:9: note: in expansion of macro ‘ENSURE’
  396 |         ENSURE(vfprintf);
      |         ^~~~~~
In file included from /usr/include/stdio.h:867,
                 from /home/j3tracey/shadow/src/preload/shd-interposer.c:12:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:98:1: error: inlining failed in call to always_inline ‘fprintf.localalias’: redefined extern inline functions are not considered for inlining
   98 | fprintf (FILE *__restrict __stream, const char *__restrict __fmt, ...)
      | ^~~~~~~
/home/j3tracey/shadow/src/preload/shd-interposer.c:62:9: note: called from here
   62 |         fprintf(stderr, "dlsym(%s): returned NULL pointer\n", funcstr); \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/j3tracey/shadow/src/preload/shd-interposer.c:69:9: note: in expansion of macro ‘SETSYM_OR_FAIL’
   69 |         SETSYM_OR_FAIL(director.next.func, #func); \
      |         ^~~~~~~~~~~~~~
/home/j3tracey/shadow/src/preload/shd-interposer.c:396:9: note: in expansion of macro ‘ENSURE’
  396 |         ENSURE(vfprintf);
      |         ^~~~~~
/home/j3tracey/shadow/src/preload/shd-interposer.c: At top level:
cc1: warning: unrecognized command line option ‘-Wno-static-in-inline’
```

**To Reproduce**
`./setup build -c` on Ubuntu 19.10 with a fresh checkout of Shadow. 

**Operating System (please complete the following information):**
 - OS and version: post the output of `lsb_release -d`  
Ubuntu 19.10
 - Kernel version: post the output of `uname -a`  
`5.3.0-40-generic #32-Ubuntu SMP Fri Jan 31 20:24:34 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`

**Shadow (please complete the following information):**
 - Version: post the output of `shadow --version`  
Basically every version, including master.
 - Which plug-ins you are using:  
N/A

**Comment by sporksmith** [comment]
Used `--verbose` in the setup script to get the compiler commands, and then changed -c to -E to preprocess shd-interposer.c:

```
$ cd /shadow/build/src/preload && /usr/bin/gcc -DNDEBUG -DVERBOSE -Dshadow_interpose_EXPORTS -I/shadow/build/src/preload -I/shadow/src/preload -I/shadow/src/main -I/shadow/build -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -I/usr/include/igraph -I/home/user/.shadow/include  -fPIC   -ggdb -fno-omit-frame-pointer -O3 -Wno-static-in-inline -o CMakeFiles/shadow-interpose.dir/shd-interposer.c.e   -E /shadow/src/preload/shd-interposer.c
```

The output has an always_inline definition of fprintf from `/usr/include/x86_64-linux-gnu/bits/stdio2.h`:

```
# 26 "/usr/include/x86_64-linux-gnu/bits/stdio2.h" 3 4
...

extern __inline __attribute__ ((__always_inline__)) __attribute__ ((__gnu_inline__)) __attribute__ ((__artificial__)) int
fprintf (FILE *__restrict __stream, const char *__restrict __fmt, ...)
{
  return __fprintf_chk (__stream, 2 - 1, __fmt,
   __builtin_va_arg_pack ());
}
```

Here's the un-preprocessed definition in that header file:
```
#if __USE_FORTIFY_LEVEL > 1
...
# ifdef __va_arg_pack
__fortify_function int
fprintf (FILE *__restrict __stream, const char *__restrict __fmt, ...)
```

It looks like glibc has a public-ish interface for disabling this, by unsetting [_FORTIFY_SOURCE](https://gcc.gnu.org/legacy-ml/gcc-patches/2004-09/msg02055.html). Adding -U_FORTIFY_SOURCE to the CFLAGS indeed gets the compilation to finish. 

```
--- a/src/preload/CMakeLists.txt
+++ b/src/preload/CMakeLists.txt
@@ -5,6 +5,7 @@ find_package(DL REQUIRED)
 include_directories(${DL_INCLUDES} ${IGRAPH_INCLUDES} ${GLIB_INCLUDES})
 
 #add_cflags(-Wno-invalid-noreturn)
+add_cflags(-U_FORTIFY_SOURCE)
 
 ## build the required library for intercepting system and library calls.
 ## create the interpose library, to be set as LD_PRELOAD to intercept functions.
```

I'm not sure this is the best solution, but you could try it if you need a short term workaround.

**Comment by sporksmith** [comment]
An alternative solution is to not try to interpose fprintf at all, and rely on interposing lower level IO functionality that it calls.

[Experimentally](https://godbolt.org/z/U3NpZY), fprintf gets compiled into a call to fwrite, which we also interpose. If that's the case then getting rid of the fprintf wrapper should be fine.

The risk is if in some cases it instead gets compiled to some internal glibc IO function that we *don't* interpose, but it seems relatively unlikely.

**Comment by sporksmith** [comment]
I suspect the most robust solution is to avoid including the system libc headers that declare any symbols that we're redefining in that compilation unit, since in general there's no guarantee that the headers won't do things like this. e.g. our implementation of `fprintf` should be in a compilation unit that doesn't include `<stdio.h>`.

There's WIP to reimplement the shim-side functionality. I think it makes sense to apply that principle there, and leave this as-is for the moment. If someone needs a short term workaround in the meantime, here are the collection mentioned above:

- Use Ubuntu 18 instead (the platform we target for continuous integration).
- Use clang instead of gcc
- Apply the patch [above](https://github.com/shadow/shadow/issues/737#issuecomment-597162972) to undefine `_FORTIFY_SOURCE`

**Comment by jtracey** [comment]
Debug builds (`-g`) are also a workaround for now (albeit with slightly slower performance and higher memory overhead). The reason I didn't notice this sooner is only my dev machine is on 19.10, and I never have a reason to build Shadow without debug symbols there. :)

**Comment by yihuajack** [comment]
> Debug builds (`-g`) are also a workaround for now (albeit with slightly slower performance and higher memory overhead). The reason I didn't notice this sooner is only my dev machine is on 19.10, and I never have a reason to build Shadow without debug symbols there. :)

It seems that to add a parameter `-g` or `--debug` does eliminate any build errors but setup test would fail.
